alpha/veneer/semver/semver.go

```diff
@@ -0,0 +1,428 @@
+	data, err := ioutil.ReadFile(ref)
+	if err != nil {
+		return nil, err
+	}
+
+	// default behavior is to generate only minor channels and to use skips over replaces
+	sv := semverVeneer{
+		GenerateMajorChannels: false,
+		GenerateMinorChannels: true,
+		AvoidSkipPatch:        false,
+	}
+	if err := yaml.Unmarshal(data, &sv); err != nil {
+		return nil, err
+	}
+	return &sv, nil
+}
+
+func (sv *semverVeneer) getVersionsFromStandardChannels(cfg *declcfg.DeclarativeConfig) (*semverRenderedChannelVersions, error) {
+	versions := semverRenderedChannelVersions{}
+
+	bdm, err := sv.getVersionsFromChannel(sv.Candidate.Bundles, cfg)
+	if err != nil {
+		return nil, err
+	}
+	versions[CandidateChannelName] = bdm
+
+	bdm, err = sv.getVersionsFromChannel(sv.Fast.Bundles, cfg)
+	if err != nil {
+		return nil, err
+	}
+	versions[FastChannelName] = bdm
+
+	bdm, err = sv.getVersionsFromChannel(sv.Stable.Bundles, cfg)
+	if err != nil {
+		return nil, err
+	}
+	versions[StableChannelName] = bdm
+
+	return &versions, nil
+}
+
+func (sv *semverVeneer) getVersionsFromChannel(semverBundles []semverVeneerBundleEntry, cfg *declcfg.DeclarativeConfig) (map[string]semver.Version, error) {
+	entries := make(map[string]semver.Version)
+
+	for _, semverBundle := range semverBundles {
+		// for some reason, not able to get the find-by-image-name to function properly with
+		// disjoint sort.Slice and sort.Search operations.
+		// for right now, replace with a sort.Search-like inline operation
+		//
+		// index := sort.Search(len(cfg.Bundles), func(i int) bool {
+		// 	return cfg.Bundles[i].Image == b.Image
+		// })
+		index := 0
+		for index < len(cfg.Bundles) {
+			if cfg.Bundles[index].Image == semverBundle.Image {
+				break
+			}
+			index++
+		}
+		if index == len(cfg.Bundles) {
+			return nil, fmt.Errorf("supplied bundle image name %q not found in rendered bundle images", semverBundle.Image)
+		}
+		b := cfg.Bundles[index]
+
+		props, err := property.Parse(b.Properties)
+		if err != nil {
+			return nil, fmt.Errorf("parse properties for bundle %q: %v", b.Name, err)
+		}
+		if len(props.Packages) != 1 {
+			return nil, fmt.Errorf("bundle %q has multiple %q properties, expected exactly 1", b.Name, property.TypePackage)
+		}
+		v, err := semver.Parse(props.Packages[0].Version)
+		if err != nil {
+			return nil, fmt.Errorf("bundle %q has invalid version %q: %v", b.Name, props.Packages[0].Version, err)
+		}
+
+		// if we have a known package name, then ensure all new packages match
+		if sv.pkg != "" {
+			if props.Packages[0].PackageName != sv.pkg {
+				return nil, fmt.Errorf("bundle %q does not belong to this package: %q", props.Packages[0].PackageName, sv.pkg)
+			}
+		} else {
+			// else cache the first
+			p := newPackage(props.Packages[0].PackageName)
+			cfg.Packages = append(cfg.Packages, *p)
+			sv.pkg = props.Packages[0].PackageName
+		}
+
+		entries[b.Name] = v
+	}
+
+	return entries, nil
+}
+
+type highwaterChannel struct {
```

Comment:
add comment/description:
highwater-mark captures here the concept of the most-stable channel-head.  Like a real hwm, it will 'float up' to the greatest of the core comparison.